### PR TITLE
batches: migrate list page, enable filtering to multiple statuses

### LINKED ISSUE #31164: batches: Rethink filters on batch change page
We recently defaulted to "All" as the standard, because we felt that Draft BCs were hidden too much. 
While I agree, I think this causes another problem, especially on the global list, where the list might be extremely long by default because it includes all those closed ones from the past.
Maybe we should default to draft + open, instead of "All", like GitHub does for PRs and issues?

### PR DESCRIPTION
**Pull request by courier-new**
Closes #31164.

Swaps out the filters on the batch changes index (list) page from a set of radio buttons to a new `MultiSelect`, which defaults to open + draft (or just open, if running batch changes server-side is not enabled).

image

Design based on [this Figma screen](https://www.figma.com/file/cTKjwbvWqU6xAykZmCTTqy/SSBC?node-id=1455%3A48950), but with a multi-select in place of the single select dropdown and without the search field.

Granted, it looks a little odd right now for users without server-side execution enabled, since there are only the two options for open and closed. But I don't think it's necessarily worth the effort and extra code to preserve the radio buttons when the flag is disabled when we'll only be supporting that case for at most a couple more releases.

Most of the code changes here, as you might notice, are actually because I migrated the list view for this page from the older `requestGraphQL` and `FilteredConnection` to Apollo, `useConnection`, and the new filtered connection component composition API as part of this work. As part of _that_ migration, I was also able to simplify the logic around if batch changes is licensed and if there are any batch changes created yet (to auto-switch to the "Getting started" tab if not) by capturing them both in a single query, independent from the connection.

## Test plan

This page is heavily covered by storybook stories as well as an integration test. I've added an additional storybook story for the namespace version of this page. I also verified manually that things were functioning as expected as thoroughly as I could.

### CODE REVIEW

File: client/web/src/enterprise/batches/list/BatchChangeListPage.tsx [hunk 1]
```diff
@@ -40,108 +48,91 @@ export interface BatchChangeListPageProps
...
+            states: filterStates,
+            first: BATCH_CHANGES_PER_PAGE_COUNT,
+            after: null,
+            viewerCanAdminister: null,
+        },
+        options: { useURL: true },
+        getConnection: result => {
+            const data = dataOrThrowErrors(result)
+            if (!namespaceID) {
+                return (data as BatchChangesResult).batchChanges
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/list/BatchChangeListPage.tsx:124]: Oh, clever. 👏

File: client/web/src/enterprise/batches/list/BatchChangeNode.tsx [hunk 2]
```diff
@@ -23,7 +23,7 @@ import { BatchChangeStatePill } from './BatchChangeStatePill'
 
 export interface BatchChangeNodeProps {
     node: ListBatchChange
-    executionEnabled: boolean
+    isExecutionEnabled: boolean
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/list/BatchChangeNode.tsx:26]: I know I'm totally guilty of this a lot of the time, but I think we should try to standardise more on `is` as a prefix for state booleans. :+1:

File: client/web/src/enterprise/batches/list/backend.ts [hunk 3]
```diff
@@ -36,152 +35,63 @@ const listBatchChangeFragment = gql`
...
-            if (data.node.__typename !== 'Org' && data.node.__typename !== 'User') {
-                throw new Error(`Requested node is a ${data.node.__typename}, not a User or Org`)
-            }
-            return {
-                batchChanges: data.node.batchChanges,
-                totalCount: data.node.allBatchChanges.totalCount,
-            }
-        })
-    )
+    ${listBatchChangeFragment}
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/list/backend.ts:86]: The fragment changes here are long overdue. :+1:

File: client/web/src/enterprise/batches/list/BatchChangeListPage.tsx [hunk 4]
```diff
@@ -151,31 +142,56 @@ export const BatchChangeListPage: React.FunctionComponent<BatchChangeListPagePro
...
-                        withCenteredSummary={true}
-                        cursorPaging={true}
-                        noSummaryIfAllNodesVisible={true}
-                        emptyElement={<BatchChangeListEmptyElement canCreate={canCreate} location={location} />}
-                    />
+                    <ConnectionContainer>
+                        {!!connection?.nodes.length && (
+                            <div className="d-flex align-items-center justify-content-end mb-4">
+                                <h3 className="align-self-end flex-1">{connection?.totalCount} batch changes</h3>
+                                <h4 className="mb-0 mr-2">Status</h4>
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/list/BatchChangeListPage.tsx:154]: This needs either a left margin or a style rule to move onto a new line (probably the latter, based on the multi select behaviour) on small screens:

image

**Review by LawnGnome — APPROVED** [review]
OK, code wise this looks awesome. Nice work! There are a couple of issues on small screens, which I'm fairly sanguine if we don't address (given the awful overall state of Sourcegraph and Batch Changes on mobile), but if there are quick wins there we might as well take them.

> But I don't think it's necessarily worth the effort and extra code to preserve the radio buttons when the flag is disabled when we'll only be supporting that case for at most a couple more releases.

I tend to agree, but I guess my lingering concern here is whether it makes sense to expose draft batch changes for customers who aren't using server side changes. It _probably_ isn't worth the effort even then — it's functional; just slightly confusing — but I guess it comes down to what our vision is for Batch Changes in the longer term. Are drafts going to be exposed in the UI generally somehow if executors aren't in use?

So, yeah, on balance I'd leave it, but I guess I do want to hear if anyone else feels more strongly about this (in either direction).

File: client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss [hunk 5]
```diff
@@ -24,3 +24,8 @@
 .narrow {
     grid-template-columns: [state] 70px [info] minmax(auto, 1fr) [open] min-content [closed] min-content [merged] min-content [end];
 }
+
+.status-dropdown {
+    margin: 0;
+    width: 17rem;
```
Comment (LawnGnome) [inline client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss:30]: I don't think we need this `width`, and having it in place causes issues on smaller displays. Excluding it still looks OK to me:

image

image

image

File: client/web/src/enterprise/batches/list/BatchChangeListPage.tsx (hunk 4 above)
Comment (eseliger) [reply client/web/src/enterprise/batches/list/BatchChangeListPage.tsx:154]: +1 for new line

File: client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss (hunk 5 above)
Comment (eseliger) [reply client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss:30]: I kinda like the consistent width, but maybe we should allow the element to shrink properly when there's less than 17rem available.

**Review by eseliger — APPROVED** [review]

**Comment by eseliger** [comment]
I am totally ok with always showing the multi-select. The only reason we don't show draft right now is because we expect instances without SSBC to never have a draft batch change, so it doesn't become confusing.

File: client/web/src/enterprise/batches/list/BatchChangeNode.tsx (hunk 2 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/list/BatchChangeNode.tsx:26]: Yeah I know I haven't always been good about this either, but I'm trying to make a conscious effort to standardi(z/s)e when I see them, now! 😁

File: client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss (hunk 5 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss:30]: Yeah I think the consistent width might look better once we add a text search box beside it, but I'm not too attached to it. @rrhyne do you have any opinion on fixed vs. dynamic width for this? 🙂

In either case I can definitely make sure it shrinks properly.

File: client/web/src/enterprise/batches/list/BatchChangeListPage.tsx (hunk 4 above)
Comment (courier-new) [reply client/web/src/enterprise/batches/list/BatchChangeListPage.tsx:154]: Thank you for holding me accountable to my decisions on small screens haha. 🙇‍♀️

**Comment by courier-new** [comment]
I think in the longterm it probably makes sense to expose drafts in the UI even to instances without executors and server-side execution enabled. It facilitates collaboration on a spec file without having to pass versions of the file around your team manually or check them into a repo (and the same for other files, once we implement that ability). I think it'll become equally confusing if we talk about drafts in our docs and always have to qualify them with "if you have server-side execution enabled...", even if the benefits of having them are much smaller to a normal src-cli user. But for now they will continue to live behind the feature flag with all the other goodies we're working on. 😛

**Comment by rrhyne** [comment]
> Are drafts going to be exposed in the UI generally somehow if executors aren't in use?

I agree with Kelli here. The behavior should be consistent.

File: client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss (hunk 5 above)
Comment (rrhyne) [reply client/web/src/enterprise/batches/list/BatchChangeListPage.module.scss:41]: I like @LawnGnome's suggestion. It looks empty without anything in it at a constant width, but I do think it should have a reasonable max width.

**Pull request merged**